### ISSUE TITLE
Azure b2c AADB2COidcLoginConfigurer

### ISSUE THREAD

**Issue by letmagnau** [microsoft/azure-spring-boot#911]
## Environment
- Spring boot starter:

  - azure-active-directory-b2c-spring-boot-starter

- OS Type: Linux
- Java version: 
 - 1.9 Java14
  
## Summary
I try to integrate b2c with Jhipster stack.
About Login it seems to work good because it create correctly ClientRegsitrations on top the flows
but about the UserInfo and ProviderDetails/COnfiguration Metadata aobut the spring Oidc standard it leave some important features

When application redirect to /logout , AADB2COidcLoginConfigurer doesnt' work , no action was raised redirecting it to https://{tenant}.b2clogin.com/{tenant}.onmicrosoft.com/{policy}/oauth2/v2.0/logout?post_logout_redirect_uri=https%3A%2F%2Fjwt.ms%2F as in dcoumentation

Also i cant pick and structure a correct redirect from Client Registration becouse, as  I said, no ProvideDetails / MEtadata COnfigurerer are stored in it , why??

ALso UserInfoEndpoit is ignored

How can we manage the issuer-uri with azure?

I see that in implementation is stored the information AADB2CLogoutSuccessHandler but how it works and how I can use correctly?
 
## Reproduce steps
Logout from app

## Expected Results

custom session was invalidated but not oidc idp login

**Comment by saragluna** [comment]
Thanks for reaching out and we'll look into it.

**Comment by letmagnau** [comment]
this is a not respected specific of openid

**Comment by letmagnau** [comment]
https://openid.net/specs/openid-connect-core-1_0.html#UserInfo

**Comment by letmagnau** [comment]
any news??

**Comment by yiliuTo** [comment]
Hi @letmagnau , do you mean you hope to set customized redirect URL of logout for AAD B2C but failed? Besides, what do you mean that "in implementation is stored the information AADB2CLogoutSuccessHandler"?

**Comment by moarychan** [comment]
Hi @letmagnau , I just tried the latest sample project, it works well. The logout-success-url item must be configured, then will trigger the redirect action.
![image](https://user-images.githubusercontent.com/19926623/87033347-a7e49300-c218-11ea-8cee-5643843a4f24.png)
You can also find snippet log as follows,

```java
2020-07-09 19:26:14.279 DEBUG 6380 --- [nio-8443-exec-8] o.s.s.w.u.matcher.AntPathRequestMatcher  : Checking match of request : '/logout'; against '/logout'
2020-07-09 19:26:14.279 DEBUG 6380 --- [nio-8443-exec-8] o.s.s.w.a.logout.LogoutFilter            : Logging out user 'org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken@93efbc26: Principal: Name: [Amy Chen], Granted Authorities: [[ROLE_USER, SCOPE_/, SCOPE_openid]], User Attributes: [{at_hash=Vr1RMUa5UjK2yqaMlvuudA, sub=548e9855-41ce-42f3-9d86-86f2a045f65f, country=China, ver=1.0, postalCode=7100000, iss=https://moaryb2c.b2clogin.com/a7e59e27-a77f-4745-a47f-86243dba82d7/v2.0/, nonce=4CvDe4ejMlm3M5ofxaNFAxEeK3O0GpDYa-aECQoJ8BE, aud=[116b0d71-53f6-4951-87d5-c9e6ec347764], nbf=Thu Jul 09 19:26:08 CST 2020, auth_time=2020-07-09T11:26:08Z, name=Amy Chen, exp=2020-07-09T12:26:08Z, iat=2020-07-09T11:26:08Z, tfp=B2C_1_signupsignin1}]; Credentials: [PROTECTED]; Authenticated: true; Details: org.springframework.security.web.authentication.WebAuthenticationDetails@380f4: RemoteIpAddress: 0:0:0:0:0:0:0:1; SessionId: 0C6AAD3ACCF918C4D05767B50E61875B; Granted Authorities: ROLE_USER, SCOPE_/, SCOPE_openid' and transferring to logout destination
2020-07-09 19:26:14.279 DEBUG 6380 --- [nio-8443-exec-8] o.s.s.w.a.l.SecurityContextLogoutHandler : Invalidating session: BC1C77A09A8A7989EF082F735E1BB808
2020-07-09 19:26:14.280 DEBUG 6380 --- [nio-8443-exec-8] c.m.a.s.a.b.AADB2CLogoutSuccessHandler   : Using default Url: https://moaryb2c.b2clogin.com/moaryb2c.onmicrosoft.com/oauth2/v2.0/logout?post_logout_redirect_uri=https%3A%2F%2Flocalhost%3A8443%2Flogin&p=B2C_1_signupsignin1
2020-07-09 19:26:14.280 DEBUG 6380 --- [nio-8443-exec-8] o.s.s.web.DefaultRedirectStrategy        : Redirecting to 'https://moaryb2c.b2clogin.com/moaryb2c.onmicrosoft.com/oauth2/v2.0/logout?post_logout_redirect_uri=https%3A%2F%2Flocalhost%3A8443%2Flogin&p=B2C_1_signupsignin1'


```
If you still cannot logout successfully, I hope you can provide your dmeo project, I will help you find the root cause faster.

**Comment by letmagnau** [comment]
Hi @moarychan , i'm trying to customize the @jhipster stack for Oauth2, but we are finding some trouble in different scenarios

First of all , as descirbed when the webapp redirect /logout , I would have expected that AADB2CLogoutSuccessHandler that is initialized in AADB2COidcLoginConfigurer would react and logout me also with the identity provider that I have choosen (google in this case) but it not work as aspected. I have to custom write a snipped like this

**Comment by moarychan** [comment]
Hi @letmagnau, I'm not very sure about your real environment and how to use it. I found the corresponding points based on local testing for your reference:
1. Configuration logout-success-url in yml file, will be executed in org.springframework.security.web.authentication.AbstractAuthenticationTargetUrlRequestHandler#handle
![sdk-911](https://user-images.githubusercontent.com/19926623/87401658-003adc80-c5ed-11ea-84d3-2d1849556dad.png)


2. Configuration logout-url in authentication page of app registrations in azure portal, My understanding is this, when logging out, logout-url is called by Azure Identify Platform(Applications must respond to this request by clearing any session that identifies the user and returning a 200 response. ), and then redirected to logout-success-url. there must be a public address accessible by azure to access. 

Ref:
Documentation: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-a-sign-out-request
Similar issues: https://github.com/MicrosoftDocs/azure-docs/issues/57608      https://github.com/MicrosoftDocs/azure-docs/issues/34577

Regarding the issue you said, I understand whether it is managed here, please verify
![sdk-911-scope-mgr](https://user-images.githubusercontent.com/19926623/87401679-092bae00-c5ed-11ea-8786-7b0cddd584bf.png)

**Comment by letmagnau** [comment]
First of all, I would to thank you for your support.

For the first problem, about the logout , it can have sense. so i'm now
redirecting to localhost and it is not accessible from internet now, i'll
try with an public url

About the second issue, in the authenticate method of
OidcAuthorizationCodeAuthenticationProvider    of spring security 5 you
create mapped authority

Collection<? extends GrantedAuthority> mappedAuthorities =
   this.authoritiesMapper.mapAuthorities(oidcUser.getAuthorities());


this will be readed from the configuration of Azure Portal and as I told
you and you have correctly answered me , it is true
[image: image.png]


[image: image.png]

also for other 2 scopes inherited from my backend api (including those of
Graph) but in my authoritiues after the login in the authenticate method I
have only

[image: image.png]

ROLE USER = jhipster role default

SCOPE_/ ???? I cant understand this (please explain me what is)

And SCOPE_openid that is the OpenID standard token

so Where I'm wrong??

Where in the Principal that spring inject I can found the scopes that I 've
configured in Azure Portal??

Thanks in advance

Piero

Il giorno mar 14 lug 2020 alle ore 10:15 M <notifications@github.com> ha
scritto:

**Comment by moarychan** [comment]
Hi @letmagnau , about SCOPE_/ I found some documentation, I think a simple understanding is a default configuration that can read all permissions dynamically, you can find more details [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#using-the-admin-consent-endpoint), hope it helps you.

**Comment by moarychan** [comment]
Hi @letmagnau , is there any other update? If not, we will close this issue tomorrow.

**Comment by letmagnau** [comment]
I've tried to replicate the logout problem but It not work as expected
My stack is spring boot with aure starter : 
this is my SecutiryConfiguration

@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled = true, securedEnabled = true)
@Import(SecurityProblemSupport.class)
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

    private final CorsFilter corsFilter;
    private final AADB2COidcLoginConfigurer configurer;
/*    @Value("${spring.security.oauth2.client.provider.oidc.issuer-uri}")
    private String issuerUri;*/

    private final JHipsterProperties jHipsterProperties;
    private final SecurityProblemSupport problemSupport;

   /* public SecurityConfiguration(CorsFilter corsFilter, JHipsterProperties jHipsterProperties, SecurityProblemSupport problemSupport) {
        this.corsFilter = corsFilter;
        this.problemSupport = problemSupport;
        this.jHipsterProperties = jHipsterProperties;
    }*/

    public SecurityConfiguration(CorsFilter corsFilter, JHipsterProperties jHipsterProperties, SecurityProblemSupport problemSupport,AADB2COidcLoginConfigurer configurer) {
        this.corsFilter = corsFilter;
        this.problemSupport = problemSupport;
        this.jHipsterProperties = jHipsterProperties;
        this.configurer = configurer;
    }

    @Override
    public void configure(WebSecurity web) {
        web.ignoring()
            .antMatchers(HttpMethod.OPTIONS, "/**")
            .antMatchers("/swagger-ui/index.html")
            .antMatchers("/test/**");
    }

    @Override
    public void configure(HttpSecurity http) throws Exception {
        // @formatter:off
        http
            .csrf()
            .csrfTokenRepository(CookieCsrfTokenRepository.withHttpOnlyFalse())
        .and()
            .addFilterBefore(corsFilter, CsrfFilter.class)
            .exceptionHandling()
                .authenticationEntryPoint(problemSupport)
                .accessDeniedHandler(problemSupport)
        .and()
            .headers()
            .contentSecurityPolicy("default-src 'self'; frame-src 'self' data:; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://storage.googleapis.com; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self' data:")
        .and()
            .referrerPolicy(ReferrerPolicyHeaderWriter.ReferrerPolicy.STRICT_ORIGIN_WHEN_CROSS_ORIGIN)
        .and()
            .featurePolicy("geolocation 'none'; midi 'none'; sync-xhr 'none'; microphone 'none'; camera 'none'; magnetometer 'none'; gyroscope 'none'; speaker 'none'; fullscreen 'self'; payment 'none'")
        .and()
            .frameOptions()
            .deny()
        .and()
            .authorizeRequests()
            .antMatchers("/api/auth-info").permitAll()
            .antMatchers("/api/**").authenticated()
            .antMatchers("/management/health").authenticated()
            .antMatchers("/management/info").permitAll()
            .antMatchers("/management/prometheus").permitAll()
            .antMatchers("/management/**").hasAuthority(AuthoritiesConstants.ADMIN)
            .and()
            .apply(configurer);

        /*.and()
            .oauth2Login()
        .and()
            .oauth2ResourceServer()
                .jwt()
                .jwtAuthenticationConverter(authenticationConverter())
                .and()
            .and()
                .oauth2Client();*/
        // @formatter:on
    }



and we have a angular app that first invalidate session 

  @PostMapping("/api/logout")
    public ResponseEntity<?> logout(HttpServletRequest request,
                                    @AuthenticationPrincipal(expression = "idToken") OidcIdToken idToken) throws ServletException {


     
        Map<String, String> logoutDetails = new HashMap<>();
        System.out.println(logoutUrl);
        logoutDetails.put("logoutUrl",logoutUrl);
        request.getSession().invalidate();


     

        return ResponseEntity.ok().body(logoutDetails);




and after try to redirect forwar to /logout 


logout(): void {
     this.authServerProvider.logout().subscribe((logout: Logout) => {
    //const logoutUrl = logout.logoutUrl;
    // const redirectUri = `${location.origin}${this.location.prepareExternalUrl('/')}`;
     //const redirectUri = `${location.origin}${logoutUrl}`;

      //  logoutUrl = logoutUrl + redirectUri;
   });
   location.href = `${location.origin}${this.location.prepareExternalUrl('/logout')}`;
  }


but it doest work calling simple /logout redirect


if I recreate the url with 

 String logoutUrl  =  AADB2CURL.getEndSessionUrl(this.tenant,this.logoutUrl, this.registration.getClientName());

and pass to the angular redirect , it works

it seems thet not understand the /logout keyword and translate it in the correct logout address


sorry for that



About the scope of the B2c we cannot undestand how we can manage the scope that we have correctly set on azure portal as
@moarychan  told us

I't very frustrating. But there are more difference with Oauth2 standard like Okta or Keylock

- no userInfo endpoint
- no scp claims in response 
- no authorities (only SCOPE_/ and SCOPE_openid)

Is there any that can help us to integrate Azure B2c correclty??

thanks in advance

**Comment by letmagnau** [comment]
If this can be more clear  , you can  see an azure b2c .netcore app 

https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/4-WebApp-your-API/4-2-B2C/TodoListService

in this app you have a Client Web and WEB api with scope defined 

a client sign in on azure b2c and can ask to run a endpoint of a WebApi giving him a scope that the WebAPi is able to recognize and passtrough.

I hope that i'm clear

**Comment by moarychan** [comment]
Hi @letmagnau , can you use teams(https://teams.microsoft.com/)? so that our communication efficiency will be higher, please find me v-moaryc@microsoft.com, or please provide your microsoft account, I will contact you.

**Comment by letmagnau** [comment]
I'm here

i've invited you

i'm copperpot_susca@outllok.com


Il giorno mer 15 lug 2020 alle ore 10:24 M <notifications@github.com> ha
scritto:

**Comment by letmagnau** [comment]
copperpot_susca@outlook.com


Il giorno mer 15 lug 2020 alle ore 10:55 letmagnau <letmagnau@gmail.com> ha
scritto:

**Comment by saragluna** [comment]
@letmagnau posting your questions here for the record:

> I want to recap my doubt only to be clear: 
> - how two registered app (client and server) can communicate with scope using spring boot?
> - Why if I use the webapp with AADB2COidcLoginConfigurer with an registration client that have more scopes those will not be present in the authorities of that user?
> - When I configure the scopes on Azure Portal, I need to manage something on the rest body in order to check if  who is call the endpoint have the correct right?
> - Why UserInfo EndPoint that is present in Okta or Keyloack is not present in Azure B2c?
> - In production, i have to use the Redirect URI in the registraion app on the azure portal in order to use the app?

**Comment by saragluna** [comment]
For the first question:

> how two registered app (client and server) can communicate with scope using spring boot?

Now with our `azure-active-directory-b2c-spring-boot-starter` this could not be achieved since our starter only covers the case OIDC part, which means we set the scope to only `openid`. But you could add a customized `ClientRegistration` with more scopes, e.g. `https://contoso.onmicrosoft.com/api/read openid offline_access`. Please also check this: https://docs.microsoft.com/en-us/azure/active-directory-b2c/access-tokens#scopes.

**Comment by saragluna** [comment]
For the second question, 

> Why if I use the webapp with AADB2COidcLoginConfigurer with an registration client that have more scopes those will not be present in the authorities of that user?

How did you create the `ClientRegistration`? It should be like this:

```java

return ClientRegistration.withRegistrationId(userFlow) // Use flow as registration Id.
                ....
                .scope("openid", "https://xxx.onmicrosoft.com/api/demo.write", "https://xxx.onmicrosoft.com/api/demo.read")
                ....
                .build();
```

And then the scope will appear as an authority, check the code here: https://github.com/spring-projects/spring-security/blob/master/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/userinfo/OidcUserService.java#L136.

**Comment by saragluna** [comment]
For the third question,
> When I configure the scopes on Azure Portal, I need to manage something on the rest body in order to check if who is call the endpoint have the correct right?

I don't get this question. What did you mean by rest body?

**Comment by saragluna** [comment]
For the forth question,

> Why UserInfo EndPoint that is present in Okta or Keyloack is not present in Azure B2c?

I'm afraid that AAD B2C does not provide a UserInfo Endpoint now. Please add your feedback here: https://feedback.azure.com/forums/169401-azure-active-directory/suggestions/34008505-we-definitely-need-to-have-support-for-userinfo-en

**Comment by saragluna** [comment]
For the fifth question,

> In production, I have to use the Redirect URI in the registration app on the azure portal in order to use the app?

I'm afraid so.

**Comment by letmagnau** [comment]
Thanjs for support @saragluna,


about first and second question: I've already tried to implement a resource server but my question abut this are:
1) if I use 
 @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.authorizeRequests()
            .antMatchers("/api/**")
            .authenticated()
            .and()
            .oauth2ResourceServer(OAuth2ResourceServerConfigurer::jwt);
    }

I need a url for spriung security 
spring:
  security:
      oauth2:
        resourceserver:
          jwt:
            jwk-set-uri https://maggiolieolob2c.b2clogin.com/maggiolieolob2c.onmicrosoft.com/b2c_1_signupsignin/discovery/v2.0/keys

is a correct choice that link? or it could be better something managed by azure libs like flows do?


2) if you read the sample of .netcore 
   https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/4-WebApp-your-API/4-2-B2C

if you see appsetting.json

{
  _"AzureAdB2C": {
    "Instance": "https://fabrikamb2c.b2clogin.com",
    "ClientId": "90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6",
    "Domain": "fabrikamb2c.onmicrosoft.com",
    "SignedOutCallbackPath": "/signout/B2C_1_susi",
    "SignUpSignInPolicyId": "b2c_1_susi",
    "ResetPasswordPolicyId": "b2c_1_reset",
    "EditProfilePolicyId": "b2c_1_edit_profile" // Optional profile editing policy
    //"CallbackPath": "/signin/B2C_1_sign_up_in"  // defaults to /signin-oidc
  },_
  "Kestrel": {
    "Endpoints": {
      "Http": {
        "Url": "http://localhost:44351"
      }
    }
  },
  "Logging": {
    "LogLevel": {
      "Default": "Warning"
    }
  },
  "AllowedHosts": "*"
}

you will see that a refferring of tenandt and client id was passed , have the starter project will include this ?? there is a way to emulate this in a correct way?


About the last answer, i have client app registered to my tenant, it redirect to a localhost for now (dev) , but if I create a postman
request pointing the Application ID URI of the app , it's not available, my question is, is there a gateway ? 

In conclusion , we have tried to use the scope  offline_access and profile as you told , but we have an unauthorized request also 
if we have allowed that graph api in the app (with Expose API features)

best regards

Piero

**Comment by letmagnau** [comment]
sorry I have closed it for mistake

**Comment by letmagnau** [comment]
sorry I have closed it for mistake

**Comment by letmagnau** [comment]
any news ? please

**Comment by saragluna** [comment]
Sorry for the late response. 

The JWK set URI is correct, we are contacting it that way too. Check it here https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/spring/azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java#L25.

**Comment by saragluna** [comment]
About the second question, what do you mean the starter will include this?

**Comment by saragluna** [comment]
>but if I create a postman request pointing the Application ID URI of the app, it's not available, my question is, is there a gateway

I don't get it. Could you explain about what are you going to achieve here?

**Comment by letmagnau** [comment]
> Sorry for the late response.
> 
> The JWK set URI is correct, we are contacting it that way too. Check it here https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/spring/azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java#L25.

ok , but there's an official spring azure resource server implementation?

**Comment by letmagnau** [comment]
> About the second question, what do you mean the starter will include this?

if you read the appsettings.json of .net , it include definition of the client id and tenant, if I use the spring resource server, the jwt-uri would be enought, it seems very strange and It make me feel very doubtful  about my problem solution.. . do you think?

**Comment by letmagnau** [comment]
> > but if I create a postman request pointing the Application ID URI of the app, it's not available, my question is, is there a gateway
> 
> I don't get it. Could you explain more about what are you going to achieve here?

if I try to browse  https://xxxxx.onmicrosoft.com/xxxx  application uri in azure portal , it don't resolve nothing

**Comment by saragluna** [comment]
We don't have a azure resource server implementation now. I believe this project with tenantId and clientId set is because it contains implementation for a web app(which in your case is your Angular SPA) and web API endpoint.

**Comment by saragluna** [comment]
> > > but if I create a postman request pointing the Application ID URI of the app, it's not available, my question is, is there a gateway
> > 
> > 
> > I don't get it. Could you explain more about what are you going to achieve here?
> 
> if I try to browse https://xxxxx.onmicrosoft.com/xxxx application uri in azure portal , it don't resolve nothing

Are you trying to access your Web API endpoint via this path?

**Comment by letmagnau** [comment]
> > > > but if I create a postman request pointing the Application ID URI of the app, it's not available, my question is, is there a gateway
> > > 
> > > 
> > > I don't get it. Could you explain more about what are you going to achieve here?
> > 
> > 
> > if I try to browse https://xxxxx.onmicrosoft.com/xxxx application uri in azure portal , it don't resolve nothing
> 
> Are you trying to access your Web API endpoint via this path?

What do you mean? Yes I try to access the api with  that path

**Comment by saragluna** [comment]
I don't think you could access your web API endpoint via this path https://xxxxx.onmicrosoft.com/application-uri
.

**Comment by letmagnau** [comment]
But I've obviously tried to call a postman request with
https://xxxxx.onmicrosoft.com/application-uri
and it's non not available

I have to do sometihing else in order to make it available?

Il giorno gio 23 lug 2020 alle ore 02:50 Xiaolu Dai <
notifications@github.com> ha scritto:

**Comment by saragluna** [comment]
Sorry for didn't make myself clear, I mean we can't access the web API endpoint via this URL https://xxxxx.onmicrosoft.com/application-uri.

**Comment by letmagnau** [comment]
and why? what it serve?

**Comment by letmagnau** [comment]
ok , and how  I can access it?

and also, what is the meaning of the application uri if i can't use it?

Il giorno lun 27 lug 2020 alle ore 02:37 Xiaolu Dai <
notifications@github.com> ha scritto:

**Comment by saragluna** [comment]
In my understanding, the application id URI is used for authentication. And if you want to redirect to your service domain via this URI, I don't know how to set this up (or whether AAD supports such function). Did you happen to read some material about this scenario? You could correct me if I am wrong about this.

**Comment by chenrujun** [comment]
Closing this issue.
Because it's not active for a long time.
If anyone have similar issue, please create issue in new repo: https://github.com/Azure/azure-sdk-for-java/issues
